File: compiler/archflow/validator.py
from __future__ import annotations

from collections import defaultdict

from .model import ArchFlowAst, Diagnostic
from .runtime import validate_runtime


def validate_ast(ast: ArchFlowAst) -> list[Diagnostic]:
    diagnostics: list[Diagnostic] = []
    diagnostics.extend(_validate_components(ast))
    diagnostics.extend(_validate_channels(ast))
    diagnostics.extend(_validate_workflows(ast))
    diagnostics.extend(_validate_cycles(ast))
    diagnostics.extend(validate_runtime(ast))
    return diagnostics


def _validate_components(ast: ArchFlowAst) -> list[Diagnostic]:
    diagnostics: list[Diagnostic] = []
    for component in ast.components.values():
        if not component.explicit:
            diagnostics.append(
                Diagnostic(
                    "warning",
                    f"Component @{component.name} is only declared as a channel endpoint and has no module block",
                    component.line,
                )
            )
            continue
        if "Stack" not in component.attributes:
            diagnostics.append(Diagnostic("warning", f"Component @{component.name} has no .Stack attribute", component.line))
    return diagnostics


def _validate_channels(ast: ArchFlowAst) -> list[Diagnostic]:
    diagnostics: list[Diagnostic] = []
    seen: set[tuple[str, str, str]] = set()
    for channel in ast.channels:
        key = (channel.source, channel.via, channel.target)
        if key in seen:
            diagnostics.append(Diagnostic("warning", f"Duplicate channel @{channel.source} => [{channel.via}] => @{channel.target}", channel.line))
        seen.add(key)

        if channel.source not in ast.components:
            diagnostics.append(Diagnostic("error", f"Channel source @{channel.source} is not defined", channel.line))
        if channel.target not in ast.components:
            diagnostics.append(Diagnostic("error", f"Channel target @{channel.target} is not defined", channel.line))
        if not channel.schema:
            diagnostics.append(Diagnostic("warning", f"Channel [{channel.via}] has no .Schema.{channel.via} contract", channel.line))
    return diagnostics


def _validate_workflows(ast: ArchFlowAst) -> list[Diagnostic]:
    diagnostics: list[Diagnostic] = []
    for component in ast.components.values():
        workflow_names: set[str] = set()
        for workflow in component.workflows:
            if workflow.name in workflow_names:
                diagnostics.append(Diagnostic("warning", f"Duplicate workflow ${workflow.name} in @{component.name}", workflow.line))
            workflow_names.add(workflow.name)
            if not workflow.sequence:
                diagnostics.append(Diagnostic("warning", f"Workflow ${workflow.name} in @{component.name} has no main sequence", workflow.line))
    return diagnostics


def _validate_cycles(ast: ArchFlowAst) -> list[Diagnostic]:
    graph: dict[str, list[str]] = defaultdict(list)
    line_by_edge: dict[tuple[str, str], int | None] = {}
    for channel in ast.channels:
        graph[channel.source].append(channel.target)
        line_by_edge[(channel.source, channel.target)] = channel.line

    visited: set[str] = set()
    active: set[str] = set()
    diagnostics: list[Diagnostic] = []

    def visit(node: str, path: list[str]) -> None:
        if node in active:
            cycle = path[path.index(node) :] + [node]
            diagnostics.append(Diagnostic("warning", f"Channel graph contains a cycle: {' -> '.join(cycle)}", line_by_edge.get((path[-1], node))))
            return
        if node in visited:
            return
        active.add(node)
        for child in graph.get(node, []):
            visit(child, path + [child])
        active.remove(node)
        visited.add(node)

    for component_name in ast.components:
        visit(component_name, [component_name])
    return diagnostics

File: playground/src/compiler/validator.ts
import type { ArchFlowAst, Diagnostic } from "./types";

export function validateArchFlow(ast: ArchFlowAst): Diagnostic[] {
  return [
    ...validateComponents(ast),
    ...validateChannels(ast),
    ...validateWorkflows(ast),
    ...validateCycles(ast),
    ...validateRuntime(ast),
  ];
}

function validateComponents(ast: ArchFlowAst): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];
  Object.values(ast.components).forEach((component) => {
    if (!component.explicit) {
      diagnostics.push({
        severity: "warning",
        message: `Component @${component.name} is only declared as a channel endpoint and has no module block`,
        line: component.line,
      });
      return;
    }
    if (!component.attributes.Stack) {
      diagnostics.push({
        severity: "warning",
        message: `Component @${component.name} has no .Stack attribute`,
        line: component.line,
      });
    }
  });
  return diagnostics;
}

function validateChannels(ast: ArchFlowAst): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];
  const seen = new Set<string>();
  ast.channels.forEach((channel) => {
    const key = `${channel.source}\0${channel.via}\0${channel.target}`;
    if (seen.has(key)) {
      diagnostics.push({
        severity: "warning",
        message: `Duplicate channel @${channel.source} => [${channel.via}] => @${channel.target}`,
        line: channel.line,
      });
    }
    seen.add(key);

    if (!ast.components[channel.source]) {
      diagnostics.push({ severity: "error", message: `Channel source @${channel.source} is not defined`, line: channel.line });
    }
    if (!ast.components[channel.target]) {
      diagnostics.push({ severity: "error", message: `Channel target @${channel.target} is not defined`, line: channel.line });
    }
    if (!channel.schema) {
      diagnostics.push({
        severity: "warning",
        message: `Channel [${channel.via}] has no .Schema.${channel.via} contract`,
        line: channel.line,
      });
    }
  });
  return diagnostics;
}

function validateWorkflows(ast: ArchFlowAst): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];
  Object.values(ast.components).forEach((component) => {
    const workflowNames = new Set<string>();
    component.workflows.forEach((workflow) => {
      if (workflowNames.has(workflow.name)) {
        diagnostics.push({
          severity: "warning",
          message: `Duplicate workflow $${workflow.name} in @${component.name}`,
          line: workflow.line,
        });
      }
      workflowNames.add(workflow.name);
      if (workflow.sequence.length === 0) {
        diagnostics.push({
          severity: "warning",
          message: `Workflow $${workflow.name} in @${component.name} has no main sequence`,
          line: workflow.line,
        });
      }
    });
  });
  return diagnostics;
}

function validateCycles(ast: ArchFlowAst): Diagnostic[] {
  const graph = new Map<string, string[]>();
  const lineByEdge = new Map<string, number | undefined>();
  ast.channels.forEach((channel) => {
    graph.set(channel.source, [...(graph.get(channel.source) ?? []), channel.target]);
    lineByEdge.set(edgeKey(channel.source, channel.target), channel.line);
  });

  const diagnostics: Diagnostic[] = [];
  const visited = new Set<string>();
  const active = new Set<string>();

  function visit(node: string, path: string[]): void {
    if (active.has(node)) {
      const cycle = [...path.slice(path.indexOf(node)), node];
      diagnostics.push({
        severity: "warning",
        message: `Channel graph contains a cycle: ${cycle.join(" -> ")}`,
        line: lineByEdge.get(edgeKey(path.at(-1) ?? "", node)),
      });
      return;
    }
    if (visited.has(node)) {
      return;
    }
    active.add(node);
    (graph.get(node) ?? []).forEach((child) => visit(child, [...path, child]));
    active.delete(node);
    visited.add(node);
  }

  Object.keys(ast.components).forEach((componentName) => visit(componentName, [componentName]));
  return diagnostics;
}

function validateRuntime(ast: ArchFlowAst): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];
  const inboundByComponent = new Map<string, Set<string>>();
  const outboundByComponent = new Map<string, Set<string>>();

  ast.channels.forEach((channel) => {
    inboundByComponent.set(channel.target, new Set([...(inboundByComponent.get(channel.target) ?? []), channel.via]));
    outboundByComponent.set(channel.source, new Set([...(outboundByComponent.get(channel.source) ?? []), channel.via]));
  });

  Object.values(ast.components).forEach((component) => {
    Object.keys(component.attributes).forEach((key) => {
      if (key.startsWith("Expose.")) {
        const channelName = key.slice("Expose.".length);
        if (!inboundByComponent.get(component.name)?.has(channelName)) {
          diagnostics.push({
            severity: "warning",
            message: `Component @${component.name} exposes [${channelName}] but has no inbound channel with that name`,
            line: component.line,
          });
        }
      }
      if (key.startsWith("Use.")) {
        const channelName = key.slice("Use.".length).split(".")[0];
        if (!outboundByComponent.get(component.name)?.has(channelName)) {
          diagnostics.push({
            severity: "warning",
            message: `Component @${component.name} uses [${channelName}] but has no outbound channel with that name`,
            line: component.line,
          });
        }
      }
    });
  });

  return diagnostics;
}

function edgeKey(source: string, target: string): string {
  return `${source}\0${target}`;
}
